scalac-scoverage-runtime/shared/src/main/scala/scoverage/Invoker.scala

```diff
@@ -59,6 +66,80 @@ object Invoker {
+   * needed as ahe runtime process may be running on a different host than the compile time one.
+   * However, since the instruments are on the classpath, we can retrieve the instruments from different hosts.
+   *
+   * @param id the id of the statement that was invoked
+   * @param instrumentsDir the directory where the instrument data is held and needs to be copied from
+   */
+  def invokedUseEnvironment(id: Int, instrumentsDir: String): Unit = {
+    // Generating a hash to add as a subdir to the basedir. This is because every
+    // instrumented binary has a coverage object for which measurements will be generated
+    // and it is possible that a test file tests multiple instrumented binaries. Thus, to
+    // segregate reports for each source file, we need this hash. This was not an issue with
+    // previous version as then we explicitly specified a different [dataDir] each time.
+    val instrumentHash = md5HashString(instrumentsDir)
+
+    // Get the output data directory from the environment variable or write to a subdir.
+    var dataDir: String = ""
+    sys.env.get(measurementFileEnv) match {
+      case Some(d) => dataDir = d
+      case None => dataDir = subdir
+    }
+
+    dataDir += s"/$instrumentHash"
+    // [sam] we can do this simple check to save writing out to a file.
+    // This won't work across JVMs but since there's no harm in writing out the same id multiple
+    // times since for coverage we only care about 1 or more, (it just slows things down to
+    // do it more than once), anything we can do to help is good. This helps especially with code
+    // that is executed many times quickly, eg tight loops.
+    if (!dataDirToIds.contains(dataDir)) {
+      // When code is changed in source file, old measurements present
+      // inside the dataDir can skew the results. Thus, clean the dataDir to remove
+      // old measurements.
+      resetDataDir(dataDir)
+      // copy instruments.
+      if (Files.exists(Paths.get(s"$instrumentsDir/$CoverageFileName"))) {
+        copyCoverageFile(instrumentsDir, dataDir)
+      }
+      // Guard against SI-7943: "TrieMap method getOrElseUpdate is not thread-safe".
+      dataDirToIds.synchronized {
+        if (!dataDirToIds.contains(dataDir)) {
+          dataDirToIds(dataDir) = ThreadSafeMap.empty[Int, Any]
+        }
+      }
+    }
+    val ids = dataDirToIds(dataDir)
+    if (!ids.contains(id)) {
+      // Each thread writes to a separate measurement file, to reduce contention
+      // and because file appends via FileWriter are not atomic on Windows.
+      var files = threadFiles.get()
+      if (files == null) {
+        files = mutable.HashMap.empty[String, FileWriter]
+        threadFiles.set(files)
+      }
+      val writer = files.getOrElseUpdate(dataDir, new FileWriter(measurementFile(dataDir), true))
+      writer.append(Integer.toString(id)).append("\n").flush()
```

Comment:
What is this ID used for?

Reply:
Each statement is given an Id. Here Id refers to the statements which gets invoked